### PR TITLE
Replace type filter in the LogViewer by filter groups

### PR DESCRIPTION
**Pull request by alexfdezsauco**
Fixes #ORCOMP-372.

### Changes proposed in this pull request:

- Replace type filter in the LogViewer by filter groups

### CODE REVIEW

File: src/Orc.Controls/Controls/LogViewer/Models/Action.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Action.cs" company="WildGums">
+//   Copyright (c) 2008 - 2018 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Orc.Controls.Models
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Models/Action.cs]: Please use Orc.Controls namespace and give it a more meaningful name (e.g. LogFilterAction)

File: src/Orc.Controls/Controls/LogViewer/Models/ExpressionType.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="ExpressionType.cs" company="WildGums">
+//   Copyright (c) 2008 - 2018 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Orc.Controls.Models
+{
+    public enum ExpressionType
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Models/ExpressionType.cs]: Same here, use a better name (e.g. LogFilterExpressionType) and Orc.Controls namespace.

File: src/Orc.Controls/Controls/LogViewer/Models/LogFilter.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="LogFilter.cs" company="WildGums">
+//   Copyright (c) 2008 - 2018 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Orc.Controls.Models
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Models/LogFilter.cs]: Orc.Controls namespace

File: src/Orc.Controls/Controls/LogViewer/Models/LogFilter.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+        public bool Pass(LogEntry logEntry)
+        {
+            var result = false;
+            var expression = Target == Target.TypeName ? logEntry.Log.TargetType.FullName : logEntry.Log.TargetType.Assembly.GetName().Name;
+            if (!string.IsNullOrWhiteSpace(expression))
+            {
+                switch (ExpressionType)
+                {
+                    case ExpressionType.Contains:
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Models/LogFilter.cs]: Use newlines between the case statements

File: src/Orc.Controls/Controls/LogViewer/Models/LogFilterGroup.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="LogFilterGroup.cs" company="WildGums">
+//   Copyright (c) 2008 - 2018 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace Orc.Controls.Models
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Models/LogFilterGroup.cs]: Orc.Controls namespace

File: src/Orc.Controls/Controls/LogViewer/Models/Target.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Target.cs" company="WildGums">
+//   Copyright (c) 2008 - 2018 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Orc.Controls.Models
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Models/Target.cs]: Orc.Controls namespace

File: src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="ApplicationFilterGroupService.cs" company="WildGums">
+//   Copyright (c) 2008 - 2018 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace Orc.Controls.Services
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs]: Orc.Controls namespace

File: src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            Argument.IsNotNull(() => xmlSerializer);
+
+            _xmlSerializer = xmlSerializer;
+        }
+
+        public IEnumerable<LogFilterGroup> Load()
+        {
+            var applicationDataDirectory = Catel.IO.Path.GetApplicationDataDirectory();
+            var configFile = Path.Combine(applicationDataDirectory, LogFilterGroupsConfigFile);
+            if (File.Exists(configFile))
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs]: Use IFileService instead

File: src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            {
+                try
+                {
+                    using (var stream = File.OpenRead(configFile))
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs]: IFileService.OpenRead

File: src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        private readonly IXmlSerializer _xmlSerializer;
+
+        public ApplicationFilterGroupService(IXmlSerializer xmlSerializer)
+        {
+            Argument.IsNotNull(() => xmlSerializer);
+
+            _xmlSerializer = xmlSerializer;
+        }
+
+        public IEnumerable<LogFilterGroup> Load()
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs]: Make async (e.g. LoadAsync)

File: src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+                catch (Exception ex)
+                {
+                    Log.Error(ex);
+                }
+            }
+
+            return ArrayShim.Empty<LogFilterGroup>();
+        }
+
+        public void Save(IEnumerable<LogFilterGroup> filterGroups)
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Services/ApplicationFilterGroupService.cs]: SaveAsync

File: src/Orc.Controls/Controls/LogViewer/Services/Interfaces/IApplicationFilterGroupService.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="IApplicationFilterGroupService.cs" company="WildGums">
+//   Copyright (c) 2008 - 2018 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace Orc.Controls.Services
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Services/Interfaces/IApplicationFilterGroupService.cs]: Orc.Controls

File: src/Orc.Controls/Controls/LogViewer/ViewModels/LogFilterEditorViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        public LogFilterEditorViewModel(LogFilter logFilter)
+        {
+            LogFilter = logFilter;
+        }
+
+        [Model]
+        public LogFilter LogFilter { get; set; }
+
+        [ViewModelToModel(nameof(LogFilter))]
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/ViewModels/LogFilterEditorViewModel.cs]: Use Expose instead if we don't need it in code.

File: src/Orc.Controls/Controls/LogViewer/ViewModels/LogFilterGroupEditorViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    using Catel.Data;
+    using Catel.MVVM;
+    using Catel.Services;
+    using Orc.Controls.Models;
+
+    public class LogFilterGroupEditorViewModel : ViewModelBase
+    {
+        private readonly IMessageService _messageService;
+
+        public LogFilterGroupEditorViewModel(IMessageService messageService) : this(new LogFilterGroup(), messageService)
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/ViewModels/LogFilterGroupEditorViewModel.cs]: call to base (e.g. : this()) should be on the next line.

File: src/Orc.Controls/Controls/LogViewer/ViewModels/LogFilterGroupEditorViewModel.cs [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        public LogFilterGroupEditorViewModel(LogFilterGroup logFilterGroup, IMessageService messageService)
+        {
+            Argument.IsNotNull(() => messageService);
+
+            _messageService = messageService;
+            LogFilterGroup = logFilterGroup;
+
+            AddOrSaveCommand = new Command(OnAddOrSaveCommandExecute, OnAddOrSaveCommandCanExecute);
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/ViewModels/LogFilterGroupEditorViewModel.cs]: Make it TaskCommand instead

File: src/Orc.Controls/Controls/LogViewer/ViewModels/LogFilterGroupListControlViewModel.cs [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        {
+            Argument.IsNotNull(() => applicationFilterGroupService);
+            Argument.IsNotNull(() => messageService);
+            Argument.IsNotNull(() => uIVisualizerService);
+
+            _applicationFilterGroupService = applicationFilterGroupService;
+            _messageService = messageService;
+            _uIVisualizerService = uIVisualizerService;
+
+            AddCommand = new TaskCommand(OnAddCommandExecute);
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/ViewModels/LogFilterGroupListControlViewModel.cs]: TaskCommand

File: src/Orc.Controls/Controls/LogViewer/ViewModels/LogViewerViewModel.cs [hunk 17]
```diff
@@ -373,12 +384,18 @@ private void UpdateEntriesCount(List<LogEntry> entries)
 
         private bool PassFilters(LogEntry logEntry)
         {
-            return PassTypeFilter(logEntry) && PassLogFilter(logEntry);
+            return PassTypeFilter(logEntry) && PassLogFilter(logEntry) && PassApplicationFilterGroupsConfiguration(logEntry);
+        }
+
+        private bool PassApplicationFilterGroupsConfiguration(LogEntry logEntry)
+        {
+            var enabledFilterGroups = _applicationFilterGroups.Where(filterGroup => filterGroup.IsEnabled).ToList();
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/ViewModels/LogViewerViewModel.cs]: Can we cache this somehow?

File: src/Orc.Controls/Controls/LogViewer/Views/LogFilterEditorControl.xaml [hunk 18]
```diff
@@ -0,0 +1,31 @@
+﻿<caltel:UserControl
+    xmlns:caltel="http://schemas.catelproject.com" x:Class="Orc.Controls.Views.LogFilterEditorControl"
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Views/LogFilterEditorControl.xaml]: Fix catel namespace

File: src/Orc.Controls/Controls/LogViewer/Views/LogFilterEditorControl.xaml [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+            <ColumnDefinition Width="*"/>
+        </Grid.ColumnDefinitions>
+        <Grid.RowDefinitions>
+            <RowDefinition Height="Auto"/>
+            <RowDefinition Height="Auto"/>
+            <RowDefinition Height="Auto"/>
+            <RowDefinition Height="Auto"/>
+            <RowDefinition Height="Auto"/>
+        </Grid.RowDefinitions>
+        <TextBlock Margin="5" Grid.Column="0" Grid.Row="0">Filter Name</TextBlock>
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Views/LogFilterEditorControl.xaml]: Remove all the margins and use orc:StackGrid instead

File: src/Orc.Controls/Controls/LogViewer/Views/LogFilterEditorControl.xaml.cs [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+    using System.Windows;
+    using Catel.MVVM.Views;
+    using Orc.Controls.Models;
+
+    /// <summary>
+    /// Interaction logic for LogFilterEditorControl.xaml
+    /// </summary>
+    public partial class LogFilterEditorControl
+    {
+        public static readonly DependencyProperty LogFilterProperty = DependencyProperty.Register(
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Views/LogFilterEditorControl.xaml.cs]: Keep dependency property registration & property wrapper next to eachother

File: src/Orc.Controls/Controls/LogViewer/Views/LogFilterGroupEditorWindow.xaml [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+            <ColumnDefinition Width="*"/>
+            <ColumnDefinition Width="Auto"/>
+        </Grid.ColumnDefinitions>
+        <Grid.RowDefinitions>
+            <RowDefinition Height="Auto"/>
+            <RowDefinition Height="Auto"/>
+            <RowDefinition Height="*"/>
+        </Grid.RowDefinitions>
+
+        <TextBlock Margin="5" Grid.Column="0">Group Name</TextBlock>
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Views/LogFilterGroupEditorWindow.xaml]: Remove margins and use orc:StackGrid

File: src/Orc.Controls/Controls/LogViewer/Views/LogViewerControl.xaml.cs [hunk 22]
```diff
@@ -23,6 +23,7 @@ namespace Orc.Controls
     using Catel.MVVM.Views;
     using Logging;
     using ViewModels;
+    using Action = System.Action;
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Controls/LogViewer/Views/LogViewerControl.xaml.cs]: We should use a better name so we don't have class name clashes.

File: src/Orc.Controls/Orc.Controls.csproj [hunk 23]
```diff
@@ -33,6 +33,16 @@
   </ItemGroup>
 
   <ItemGroup>
+    <Reference Include="System.Windows.Forms" />
+  </ItemGroup>
+
+  <ItemGroup>
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Orc.Controls.csproj]: Remove this code, it's auto included in the new project structure

File: src/Orc.Controls.Example.NET/ViewModels/LogViewerViewModel.cs [hunk 24]
```diff
@@ -30,6 +31,8 @@ public LogViewerViewModel()
 
         public Command AddLogRecords { get; set; }
 
+        public LogFilterGroupListControlViewModel LogFilterGroupListViewModel { get; set; }
```
[the remark this answers is not in the document]
Comment (alexfdezsauco) [reply src/Orc.Controls.Example.NET/ViewModels/LogViewerViewModel.cs]: Sorry, this actually was never used.

**Comment by alexfdezsauco** [comment]
fixed

**Pull request merged**